core/execution/future/market.go

```diff
@@ -352,6 +356,60 @@ func (m *Market) OnEpochRestore(ctx context.Context, epoch types.Epoch) {
 	// TODO karel - implement
 }
 
+func (m *Market) onEpochEndPartiesStats() {
+	if m.markPrice == nil {
+		// no mark price yet, so no reason to calculate any of those
+		return
+	}
+
+	if m.stats == nil {
+		m.stats = &types.MarketStats{}
+	}
+
+	m.stats.PartiesOpenNotionalVolume = map[string]*num.Uint{}
+	m.stats.PartiesTotalTradeVolume = map[string]*num.Uint{}
+
+	assetQuantum, err := m.collateral.GetAssetQuantum(m.settlementAsset)
+	if err != nil {
+		m.log.Panic("couldn't get quantum for asset",
+			logging.MarketID(m.mkt.ID),
+			logging.AssetID(m.settlementAsset),
+		)
+	}
+
+	// first get the open interest per party
+	partiesOpenInterest := m.position.GetPartiesLowestOpenInterestForEpoch()
+	for p, oi := range partiesOpenInterest {
+		// volume
+		openInterestVolume := num.UintZero().Mul(num.NewUint(oi), m.markPrice)
+		// scale to position decimal
+		scaledOpenInterest := openInterestVolume.ToDecimal().Div(m.positionFactor)
+		// apply quantum
+		m.stats.PartiesOpenNotionalVolume[p], _ = num.UintFromDecimal(
+			scaledOpenInterest.Div(assetQuantum),
+		)
+	}
+
+	// first get the open interest per party
+	partiesTradedVolume := m.position.GetPartiesTradedVolumeForEpoch()
+	for p, oi := range partiesTradedVolume {
+		// volume
+		tradedVolume := num.UintZero().Mul(num.NewUint(oi), m.markPrice)
+		// scale to position decimal
+		scaledOpenInterest := tradedVolume.ToDecimal().Div(m.positionFactor)
+		// apply quantum
+		m.stats.PartiesTotalTradeVolume[p], _ = num.UintFromDecimal(
+			scaledOpenInterest.Div(assetQuantum),
+		)
+	}
+}
+
+func (m *Market) GetPartiesStats() (stats *types.MarketStats) {
+	stats, m.stats = m.stats, &types.MarketStats{}
```

Comment:
I think all of this trickery warrants a comment as its quite confusing. I know what you're trying to do and it still take me a few seconds to verify that it's actually good.

Reply:
okok will do.

Reply:
wait I just looked at the method and there's no logic apart this line in it? the github comment might be on the wrong line.